tableaudocumentapi/datasource.py

```diff
@@ -258,3 +259,38 @@ def _get_metadata_objects(self):
     def _get_column_objects(self):
         return [_column_object_from_column_xml(self._datasourceTree, xml)
                 for xml in self._datasourceTree.findall('.//column')]
+
+    ###########
+    # folders
+    ###########
+    @property
+    def folders(self):
+        if not self._folders:
+            self._refresh_folders()
+        return self._folders
+
+    def _refresh_folders(self):
+        folders = Folder.all_folders_from_datasource(self)
+        self._folders = {f.name: f for f in folders}
+
+    def add_folder(self, name, role):
+        """ Adds a new, empty folder to the datasource and returns it.
+
+        Will fail if another folder of the same name already exists,
+        or if no valid role (dimensions or measures) is provided.
+        """
+
+        if name in self.folders.keys():
+            raise ValueError('Folder names must be unique')
+
+        folder = Folder.from_name_and_role(self, name, role)
```

Comment:
This is in datasource.py, so the `self` references the datasource. The Folder object needs a reference to the parent datasource as we've discussed in [the corresponding issue](https://github.com/tableau/document-api-python/issues/112#issuecomment-264297320). 

I'll admit that it looks a bit weird, mostly because python uses `self` as the first argument in methods. To make it easier to understand, I could replace this line with:

```
parent_datasource = self
folder = Folder.from_name_and_role(parent_datasource, name, role)
```